community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ParallelBatchImporter.java

```diff
@@ -116,286 +65,36 @@ public ParallelBatchImporter( File storeDir, FileSystemAbstraction fileSystem, P
-            {
-                executeStage( new IdMapperPreparationStage( config, idMapper, cachedNodes,
-                        badCollector, memoryUsageStats ) );
-                PrimitiveLongIterator duplicateNodeIds = badCollector.leftOverDuplicateNodesIds();
-                if ( duplicateNodeIds.hasNext() )
-                {
-                    executeStage( new DeleteDuplicateNodesStage( config, duplicateNodeIds, neoStore ) );
-                }
-            }
-            // Import relationships (unlinked), properties
-            Configuration relationshipConfig =
-                    configWithRecordsPerPageBasedBatchSize( config, neoStore.getRelationshipStore() );
-            RelationshipStage unlinkedRelationshipStage =
-                    new RelationshipStage( relationshipConfig, writeMonitor, relationships, idMapper,
-                            badCollector, inputCache, neoStore, storeUpdateMonitor );
-            neoStore.startFlushingPageCache();
-            executeStage( unlinkedRelationshipStage );
-            neoStore.stopFlushingPageCache();
-            idMapper.close();
-            idMapper = null;
-            // Link relationships together with each other, their nodes and their relationship groups
-            long availableMemory = maxMemory - totalMemoryUsageOf( nodeRelationshipCache, neoStore );
-            // This is where the nodeRelationshipCache is allocated memory.
-            // This has to happen after idMapped is released
-            nodeRelationshipCache.setHighNodeId( neoStore.getNodeStore().getHighId() );
-            NodeDegreeCountStage nodeDegreeStage = new NodeDegreeCountStage( relationshipConfig,
-                    neoStore.getRelationshipStore(), nodeRelationshipCache );
-            executeStage( nodeDegreeStage );
-
-            linkData( nodeRelationshipCache, neoStore, unlinkedRelationshipStage.getDistribution(),
-                    availableMemory );
+            store.createNew();
 
-            // Release this potentially really big piece of cached data
-            long peakMemoryUsage = totalMemoryUsageOf( nodeRelationshipCache, neoStore );
-            long highNodeId = nodeRelationshipCache.getHighNodeId();
-            nodeRelationshipCache.close();
-            nodeRelationshipCache = null;
-            // Defragment relationships groups for better performance
-            new RelationshipGroupDefragmenter( config, executionMonitor, numberArrayFactory )
-                    .run( max( maxMemory, peakMemoryUsage ), neoStore, highNodeId );
+            logic.importNodes();
+            logic.prepareIdMapper();
+            logic.importRelationships();
+            logic.calculateNodeDegrees();
 
-            // Count nodes per label and labels per node
-            try ( CountsAccessor.Updater countsUpdater = neoStore.getCountsStore().reset(
-                    neoStore.getLastCommittedTransactionId() ) )
+            int type = 0;
```

Comment:
I think we should hide this loop inside ImportLogic call that will try to import all the types using the same loop. Smt like: `logic.linkRelationships()` that will keep all the looping as internal detail

Reply:
We can't hide it, it's like this because of another feature that will be built on top of it, but not right now. We could make an additional convenience method tho.